astropy/table/table.py

```diff
@@ -4283,12 +4283,10 @@ def from_pandas(cls, dataframe, index=False, units=None):
                 nan = np.nan
                 if all(isinstance(x, string_types) or x is nan for x in data):
                     # Force any missing (null) values to b''.  Numpy will
-                    # upcast to str/unicode as needed.
-                    data[mask] = b""
-
-                    # When the numpy object array is represented as a list then
-                    # numpy initializes to the correct string or unicode type.
-                    data = np.array(list(data))
+                    # upcast to str/unicode as needed. We go via a list to
+                    # avoid replacing objects in a view of the pandas array and
+                    # to ensure numpy initializes to string or bytes correctly.
+                    data = np.array([b"" if m else d for (d, m) in zip(data, mask)])
```

Comment:
This code assumes that both data and mask can only be one-dimensional. This seemed implied by the existing code, but I'm not sure if this is true generically for `Series` objects. Obviously, if not, we can rewrite a bit differently.

Reply:
p.s. Without this one gets
```
                    # Force any missing (null) values to b''.  Numpy will
                    # upcast to str/unicode as needed.
>                   data[mask] = b""
E                   ValueError: assignment destination is read-only
```

Reply:
quoting the first line from https://pandas.pydata.org/docs/user_guide/dsintro.html#series

> [Series](https://pandas.pydata.org/docs/reference/api/pandas.Series.html#pandas.Series) is a one-dimensional labeled array

so I think it's a safe assumption ! 😄

Reply:
however, I'm concerned about the efficiency of using a listcomp+zip when all we really want to do is make sure we use our own copy instead of the original array. How about this instead ?
```suggestion
                    data = data.copy()
                    data[mask] = b""
```

Reply:
I thought about that too, but then the comment stated the reason for the list comprehension was that we also need to convert from object dtype to string or byte. So, this seemed the minimal change...

Reply:
Ok, I played a little bit around it and I don't see any simpler fix, which is a long way to say that I agree ! 😅